scripts/create-build-info-package.sh

```diff
@@ -0,0 +1,113 @@
+#
+
+#
+# This script creates a new package that stores build metadata for packages
+# built by Delphix. It takes one argument: the path where the
+# combined-packages artifacts have been downloaded.
+#
+
+. "${BASH_SOURCE%/*}/common.sh"
+
+if [[ -z "$TOP" ]]; then
+	echo "Must be run inside the git repsitory."
+	exit 1
+fi
+
+if [[ $# -ne 1 ]]; then
+	echo "Must specify the path of the combined-packages artifacts." 1>&2
+	exit 1
+fi
+
+PKG_DIR="$1"
+[[ -d "$PKG_DIR/packages" ]] ||
+	die "$PKG_DIR must be a path to combined-packages artifacts."
+
+#
+# Build info files will be installed into /var/delphix-buildinfo on the
+# Delphix appliance.
+#
+target="$TOP/build-info/var/delphix-buildinfo"
+mkdir -p "$target"
+
+#
+# Copy build metadata for each package
+#
+mkdir "$target/packages"
+cd "$PKG_DIR/packages"
+for pkg in */; do
+	pushd "$pkg" &>/dev/null
+	mkdir "$target/packages/$pkg"
+	for file in GIT_HASH BUILD_INFO PACKAGE_MIRROR_URL_MAIN PACKAGE_MIRROR_URL_SECONDARY metadata.json; do
+		[[ -f "$file" ]] && cp "$file" "$target/packages/$pkg/"
+	done
+	popd &>/dev/null
+done
+
+cp "$PKG_DIR/COMPONENTS" "$target/packages/"
+cp "$PKG_DIR/KERNEL_VERSIONS" "$target/packages/"
+
+#
+# Generate build metadata for appliance-build
+#
+mkdir "$target/appliance-build"
+cd "$TOP"
+git rev-parse HEAD >"$target/appliance-build/GIT_HASH"
+
+function check_env() {
+	#
+	# When the job is ran manually for testing purposes, we do not expect
+	# all environment to be set, so skip the env check.
+	#
+	[[ -n "$JENKINS_URL" ]] || return 0
+
+	local val="${!1}"
+	[[ -n "$val" ]] || die "check_env: $1 must be non-empty"
+	return 0
+}
+
+check_env APPLIANCE_BUILD_GIT_URL
+echo "$APPLIANCE_BUILD_GIT_URL" >"$target/appliance-build/GIT_URL"
+check_env APPLIANCE_BUILD_GIT_BRANCH
+echo "$APPLIANCE_BUILD_GIT_BRANCH" >"$target/appliance-build/GIT_BRANCH"
+check_env DELPHIX_PACKAGE_MIRROR_MAIN
+echo "$DELPHIX_PACKAGE_MIRROR_MAIN" >"$target/appliance-build/DELPHIX_PACKAGE_MIRROR_MAIN"
+check_env DELPHIX_PACKAGE_MIRROR_SECONDARY
+echo "$DELPHIX_PACKAGE_MIRROR_SECONDARY" >"$target/appliance-build/DELPHIX_PACKAGE_MIRROR_SECONDARY"
+check_env AWS_S3_OUTPUT
+echo "$AWS_S3_OUTPUT" >"$target/appliance-build/ARTIFACTS_S3_LOCATION"
+check_env DELPHIX_APPLIANCE_VERSION
+echo "$DELPHIX_APPLIANCE_VERSION" >"$target/appliance-build/DELPHIX_APPLIANCE_VERSION"
+
+#
+# Build the package
+#
+cd "$TOP/build-info"
+#
+# We include some random characters in the version string to avoid collisions
+# with other build-info packages built for different platforms but included in
+# the same upgrade image. The timestamp is insufficient here since the other
+# platforms are built in parallel.
+#
+rnd="$(uuidgen | tr -d '-' | fold -w 8 | head -n 1)"
+version="1.0.0-delphix-$(date '+%Y.%m.%d.%H.%M.%S')-$rnd"
+dch --create --package delphix-build-info -v "$version" \
+	"Automatically generated changelog entry."
+
+dpkg-buildpackage -uc -us -b
+
+mv ../delphix-build-info*deb "$PKG_DIR/packages/"
```

Comment:
How do we pull this package into the appliance? Does another package depend on this new package?

Reply:
Yeah, I have a separate change that adds this package to delphix-platform dependencies

Reply:
Do you have that change posted anywhere? Curious how that'll work..

Does this mean, the VM image built for a platform of a given build, might result in a different version of this package installed than a system upgraded to the upgrade image matching that same build? E.g. the upgrade image will have multiple versions of this package installed, such that on upgrade, the system may pull in the version from a build of a different platform (i.e. whatever verison is the latest in the upgrade image)? whereas the VM image will always pull in the version of the build-info package built for that specific platform?

Reply:
Actually.. maybe both upgrade and the initial VM image will still match.. since it looks like the `delphix-entire` package will pin the correct package version for the platform..? i.e. since we `apt-get install` the package matching the version in the platform specific `delphix-entire` package, I think my concern above is not the case...

Reply:
delphix-platform change is trivial for the reasons you mentioned: https://github.com/pzakha/delphix-platform/commit/2d90030b28d0d05de2a2ec4023668d24e5eb7ed1